// cards/index.js
module.exports = {
  CardFactory: {
    Cards: [],
    CardModel: function (power, hp, talentName, manacost, name, desc) {
      this.talentName = talentName;
      this.power = power;
      this.hp = hp;
      this.cost = manacost;
      this.id = guid();
      this.canAttack = false;
      this.name = name || "";
      this.desc = desc || "";
      this.dropped = false;
      
      this.getState = function() {
        var state = {};
        state.id = this.id;
        state.hp = this.hp;
        state.canAttack = this.canAttack;
        state.talentName = this.talentName;
        return state;
      };
    },
    getRandomCard: function() {
      return new this.CardModel(getRandomInt(1,3),getRandomInt(1,3),module.exports.TalentFactory.getRandomTalent(),getRandomInt(1,2), "Random card", "Generated by true random");
    },
    getRandomFromArray: function() {
      return this.fromModel(this.Cards[getRandomInt(0, this.Cards.length-1)]);
    },
    getCoin: function() {
      return new this.CardModel(0, 0, "", -1, "Coin", "Magic coin that gives 1 mana");
    },
    getHero: function() {
      return new this.CardModel(0, 30, "hero", -1, "Hero", "Core of your division");
    },
    fromModel: function(model) {
      return new this.CardModel(model.power, model.hp, model.talentName, model.cost, model.name, model.desc);
    }
  },
  TalentFactory: {
    Talents: [],
    getRandomTalent: function() {
      return this.Talents[getRandomInt(0, this.Talents.length-1)];
    },
  }
};

function getRandomInt(min, max)
{
  return Math.floor(Math.random() * (max - min + 1)) + min;
}


function guid() {
  function s4() {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
    s4() + '-' + s4() + s4() + s4();
}
